relay-server/src/services/server.rs

```diff
@@ -55,115 +40,204 @@ const CLIENT_HEADER_TIMEOUT: Duration = Duration::from_secs(5);
-            Ok(listener) => {
-                listener.set_nonblocking(true).ok();
-                let server = axum_server::from_tcp(listener)
-                    .http_config(http_config)
-                    .addr_incoming_config(addr_config)
-                    .handle(handle.clone());
-
-                relay_log::info!("spawning http server");
-                relay_log::info!("  listening on http://{}/", config.listen_addr());
-                relay_statsd::metric!(counter(RelayCounters::ServerStarting) += 1);
-                tokio::spawn(server.serve(app));
-            }
-            Err(err) => {
-                relay_log::error!("Failed to start the HTTP server: {err}");
-                std::process::exit(1);
+    let mut ka = TcpKeepalive::new().with_time(ka_timeout);
+    #[cfg(not(any(target_os = "openbsd", target_os = "redox", target_os = "solaris")))]
+    {
+        ka = ka.with_interval(ka_timeout);
+    }
+
+    #[cfg(not(any(
+        target_os = "openbsd",
+        target_os = "redox",
+        target_os = "solaris",
+        target_os = "windows"
+    )))]
+    {
+        ka = ka.with_retries(KEEPALIVE_RETRIES);
+    }
+
+    Some(ka)
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct KeepAliveAcceptor(Option<TcpKeepalive>);
+
+impl KeepAliveAcceptor {
+    /// Create a new acceptor that sets TCP_NODELAY and keep-alive.
+    pub fn new(config: &Config) -> Self {
+        Self(build_keepalive(config))
+    }
+}
+
+impl<S> Accept<TcpStream, S> for KeepAliveAcceptor {
+    type Stream = TcpStream;
+    type Service = S;
+    type Future = std::future::Ready<io::Result<(Self::Stream, Self::Service)>>;
+
+    fn accept(&self, stream: TcpStream, service: S) -> Self::Future {
+        if let Self(Some(ref tcp_keepalive)) = self {
+            let sock_ref = socket2::SockRef::from(&stream);
+            if let Err(e) = sock_ref.set_tcp_keepalive(tcp_keepalive) {
+                relay_log::trace!("error trying to set TCP keepalive: {e}");
             }
         }
 
-        tokio::spawn(async move {
-            let Shutdown { timeout } = Controller::shutdown_handle().notified().await;
-            relay_log::info!("Shutting down HTTP server");
+        if let Err(e) = stream.set_nodelay(true) {
+            relay_log::trace!("failed to set TCP_NODELAY: {e}");
+        }
 
-            match timeout {
-                Some(timeout) => handle.graceful_shutdown(Some(timeout)),
-                None => handle.shutdown(),
-            }
-        });
+        std::future::ready(Ok((stream, service)))
     }
 }
 
-fn create_listener(addr: SocketAddr, backlog: u32) -> io::Result<TcpListener> {
-    let socket = match addr {
-        SocketAddr::V4(_) => TcpSocket::new_v4(),
-        SocketAddr::V6(_) => TcpSocket::new_v6(),
-    }?;
-    socket.bind(addr)?;
+fn serve(listener: TcpListener, app: App, config: Arc<Config>) {
+    let handle = Handle::new();
+
+    let mut server = axum_server::from_tcp(listener)
```

Comment:
This still uses `axum-server` to keep the changes to a minimum. We can place our own accept loop at a later point if we want to remove that dependency.